## Intro

In this Notebook I will train and evaluate two different Trigram models.
More precisely I will use two different approaches for creating a Trigram model.
Approach 1 is to build a matrix that holds the true probability distribution of which character 
comes next given that the model has been given two particular preceeding characters.
We then sample from this distribution to produce the next character in the output 'name'.

The second approach is to use a Neural Net. We take in the two preceeding characters to generate the next one 
and so on. The goal is to achieve the same performance or loss using the Neural Net approach as with approach 1.
As you will see approach 1 is the perfect approach given the loss function I will use. 
It is impossible to achieve better loss with a Neural Net approach to the Trigram model.

In [1]:
import torch

## First approach

We begin by loading in all the names from the text file:

In [2]:
words = open('names_dataset.txt', 'r').read().splitlines()

A peak at some names

In [3]:
words[0:10]

['emma',
 'olivia',
 'ava',
 'isabella',
 'sophia',
 'charlotte',
 'mia',
 'amelia',
 'harper',
 'evelyn']

Splitting up the Dataset into a Train and Test split. 80/20

In [4]:
from sklearn.model_selection import train_test_split

# Split the data
train_names, test_names = train_test_split(words, test_size=0.2, random_state=42)  # 20% for testing

print("Number of Training Names:", len(train_names))
print("Number of Testing Names:", len(test_names))



Number of Training Names: 25626
Number of Testing Names: 6407


Initializing the matrix that will hold the counts for future characters,
please observe that we need 256 rows because for 27 characters the possible
permutations of 2 characters following each other is 27^2 = 27 * 27 = 729.
But I will actually use 729 - 27 = 720 rows because I will ignore all entries
that has the * character as the last because if that has happened the model has already stopped producing output.

In [5]:
# Every row in this matrix will eventually be filled with the prob-distribution of the next characters given two particular preceeding characters
N = torch.zeros((702, 27), dtype=torch.int32)
N.shape

torch.Size([702, 27])

Creating two mappings between integers and characters so I can represent every characters with an integer in computations

In [6]:
chars = sorted(list(set(''.join(words))))
chars.insert(0, '*')
stoi = {s:i for i, s in enumerate(chars)}
itos = {i:s for s,i in stoi.items()}
stoi
itos

{0: '*',
 1: 'a',
 2: 'b',
 3: 'c',
 4: 'd',
 5: 'e',
 6: 'f',
 7: 'g',
 8: 'h',
 9: 'i',
 10: 'j',
 11: 'k',
 12: 'l',
 13: 'm',
 14: 'n',
 15: 'o',
 16: 'p',
 17: 'q',
 18: 'r',
 19: 's',
 20: 't',
 21: 'u',
 22: 'v',
 23: 'w',
 24: 'x',
 25: 'y',
 26: 'z'}

Creating two mappings between pairs of characters and Integers. 
Ignoring all entries that end with *, because if that happens the model has already stopped producing output.

In [28]:
chars = sorted(list(set(''.join(words))))
chars.insert(0, '*')

twochars = []
for char1 in chars:
    for char2 in chars:
        if char2 != '*':
            twochars.append(f'{char1}{char2}')

doublettoi = {s:i for i, s in enumerate(twochars)}
itodoublet = {i:s for s,i in doublettoi.items()}


Time to populate our matrix with counts of future characters!
I will use * to signal beginning and end of string. This will also be treated as a character.

In [8]:
for w in train_names:
    chs = ['*'] + list(w) + ['*']
    for i in range(len(chs) - 2):
        doublet = chs[i] + chs[i + 1]
        ch2 = chs[i + 2]
        ix1 = doublettoi[doublet]
        ix2 = stoi[ch2]
        # print(f"{doublet}, {ch2}")
        N[ix1, ix2] += 1


In [9]:
# Adding a tiny bit of smoothing to N so we dont get any zero-probabilities:
N += 1

This code will visualize our matrix that holds the counts. Darker blue indicates more instances and white indicates no instances.
Beware that the matrix plot is very long since we have so many permutations of two characters.

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

plt.figure(figsize=(16, 420))
plt.imshow(N, cmap='Blues')
for i in range(702):
    for j in range(27):
        chstr = itodoublet[i] + ' ' + itos[j]
        plt.text(j, i, chstr, ha='center', va='bottom', color='gray')
        plt.text(j, i, N[i, j].item(), ha='center', va='top', color='gray')
plt.axis('off')

Now we will convert this Count Matrix to a Probability Matrix:

In [10]:
P = (N).float()
P.shape

torch.Size([702, 27])

Time to convert the Count-Matrix to a probability-Matrix. Now every row will contain a probability distribution based on the counts of that row.

In [11]:
P /= P.sum(1, keepdim=True)

This code will visualize the probabilites for the next character given the previous two characters. Please beware that this matrix is very long so after you've had a look at the matrix maybe clear the ouput so you don't have to scroll a lot.

In [ ]:


import matplotlib.pyplot as plt
%matplotlib inline

plt.figure(figsize=(16, 420))
plt.imshow(P, cmap='Blues')
for i in range(702):
    for j in range(27):
        chstr = itodoublet[i] + ' ' + itos[j]
        plt.text(j, i, chstr, ha='center', va='bottom', color='gray')
        plt.text(j, i, "{:.3f}".format(P[i, j].item()), ha='center', va='top', color='gray')
plt.axis('off')

Before I will generate names I will find the probability distribution of the first character in all the names. This is because for our Trigram model to be able to start generating outputs we need two input characters. We only have '*' in the beginning so we need a first character also. We can generate a random first character but this will produce far more unnatural names rather than if we sample from the true distribution of the first characters in our dataset. Therefore we perform the following step: 

In [13]:
names = train_names

# Initializing a dictionary
first_char_counts = {}

# Counting each first character
for name in names:
    if name:
        first_char = name[0].lower()
        if first_char in first_char_counts:
            first_char_counts[first_char] += 1
        else:
            first_char_counts[first_char] = 1

# Converting counts to probabilities
total_names = len(names)
first_char_probabilities = {char: count / total_names for char, count in first_char_counts.items()}

# Printing the results
print("Counts:", first_char_counts)
print("Probabilities:", first_char_probabilities)

# Plucking out the probabilities for each character
sorted_probs = sorted(first_char_probabilities.items())  
prob_values = [prob for char, prob in sorted_probs]  

# Converting to a PyTorch tensor
first_character_probs = torch.tensor(prob_values, dtype=torch.float32)

# Printing the tensor
print("Probability Tensor:", first_character_probs)

Counts: {'i': 479, 'm': 2005, 'k': 2360, 'e': 1215, 'c': 1245, 'l': 1252, 'a': 3479, 't': 1063, 'z': 751, 'n': 923, 'r': 1344, 's': 1647, 'b': 1050, 'p': 429, 'f': 334, 'j': 1910, 'v': 304, 'g': 544, 'd': 1326, 'o': 309, 'h': 720, 'y': 428, 'w': 256, 'q': 79, 'x': 110, 'u': 64}
Probabilities: {'i': 0.01869195348474206, 'm': 0.07824084913759463, 'k': 0.09209396706469991, 'e': 0.04741278389136034, 'c': 0.04858346991336923, 'l': 0.04885662998517131, 'a': 0.13576055568563178, 't': 0.04148130804651526, 'z': 0.029306173417622726, 'n': 0.036018106610473735, 'r': 0.052446733785998595, 's': 0.06427066260828845, 'b': 0.0409740107703114, 'p': 0.01674081011472723, 'f': 0.013033637711699056, 'j': 0.07453367673456646, 'v': 0.01186295168969016, 'g': 0.021228439865761336, 'd': 0.051744322172793254, 'o': 0.012058066026691641, 'h': 0.028096464528213533, 'y': 0.016701787247326932, 'w': 0.009989854054475923, 'q': 0.0030828065246234293, 'x': 0.0042925154140326235, 'u': 0.002497463513618981}
Probability Ten

Now we can sample from P to get the next character until we get *, in that case we stop generating and we break.

In [15]:
# We can generate 20 names for example:

g = torch.Generator().manual_seed(423283494)

for i in range(50):
    out = []

    # Sampling the first character:
    first_char = torch.multinomial(first_character_probs, num_samples=1)
    first_char = itos[first_char.item()]
    if first_char == '*':
        continue
    first_doublet = f'*{first_char}'
    first_doublet_index = torch.tensor(doublettoi[first_doublet])
    out.append(first_char)
    first_doublet = f'*{first_char}'

    ix = doublettoi[first_doublet]
    
    while True:

        # Pluck out the probability distribution for the next character
        p = P[ix]

        #print('ix: ', ix)
        #print('current input doublet: ', itodoublet[ix])
        # Sampling
        next_char_ix = torch.multinomial(p, num_samples=1, replacement=True, generator=g).item()

        #print('next char ix: ', next_char_ix)
        if next_char_ix == 0:
            break
        #print('Adding to out: ', itos[next_char_ix])
        out.append(itos[next_char_ix])
        next_doublet = ''.join(out[-2:])
        #print('next_doublet: ', next_doublet)

        ix = doublettoi[next_doublet]

    print('Name: ', ''.join(out))

Name:  munth
Name:  resaillindynklyn
Name:  ilendrnnia
Name:  dana
Name:  ber
Name:  ganna
Name:  julin
Name:  lormogurni
Name:  emmere
Name:  ivwrean
Name:  gadannogpjxjeryah
Name:  ley
Name:  joelleenah
Name:  louray
Name:  ricken
Name:  ela
Name:  bentrukwu
Name:  fis
Name:  riononna
Name:  urosabiq
Name:  fioroie
Name:  que
Name:  quindannyerevueld
Name:  raharic
Name:  caray
Name:  faivi
Name:  qdsynnjaziesminluwen
Name:  jen
Name:  reniekileelyn
Name:  lano
Name:  rian
Name:  re
Name:  atcjldmzbres
Name:  graelie
Name:  jercmla
Name:  berne
Name:  isczrvihuryn
Name:  qzana
Name:  fri


Let's evaluate our Trigram model on our test-set and see what our Negative Log-likehihood is. We want this value (loss) to be as low as possible. A value of zero indicates that our model is 100% confident about every character in every word showing up when they actually do. In other words it would be able to perfectly predict the names in our dataset. This can't happen in practice but in theory that is kind of what we aim for.

In [26]:
log_likelihood = 0.0
count = 0

for w in test_names:
    chs = ['*'] + list(w) + ['*']
    for i in range(len(chs) - 2):
        doublet = chs[i] + chs[i + 1]
        ch2 = chs[i + 2]
        ix1 = doublettoi[doublet]
        ix2 = stoi[ch2]
        prob = P[ix1, ix2]
        logprob = torch.log(prob)
        log_likelihood += logprob
        count += 1


print(f"Log Likelihood is: {log_likelihood}")
nll = -log_likelihood
print(f"Negative Log Likelihood is: {nll}")
mean_nll = nll/count
print(f"Average Negative Log Likelihood is: {mean_nll}")

Log Likelihood is: -83431.7265625
Negative Log Likelihood is: 83431.7265625
Average Negative Log Likelihood is: 2.126082420349121


So the loss we achieve on the first approach using a probability matrix is roughly 2.126.

Now we will do the same thing but with a Neural Net instead:

In [17]:
# Creating a Neural Net to solve the same task:

# Creating the Train Dataset:

xs, ys = [], []

for w in train_names:
    chs = ['*'] + list(w) + ['*']
    for i in range(len(chs) - 2):
        doublet = chs[i] + chs[i + 1]
        ch2 = chs[i + 2]
        ix1 = doublettoi[doublet]
        ix2 = stoi[ch2]
        xs.append(ix1)
        ys.append(ix2)

xs = torch.tensor(xs)
ys = torch.tensor(ys)

num_examples = xs.nelement()
print('Number of examples: ', num_examples)
xs, ys

Number of examples:  156871


(tensor([  8, 247, 367,  ..., 472, 141, 323]),
 tensor([14,  4,  9,  ..., 12, 12,  0]))

Our Neurons will each receive 702 inputs and we will have 27 neurons in total producing an output with dim 1 x 27. We will interpret these outputs as a probability distribution over the next character given an input of a two-character string.

In [18]:
# Initializing the Neural Network:

g = torch.Generator().manual_seed(2147483647)
W = torch.randn((27, 702), generator=g, requires_grad=True)

This is the code that will process inputs with our neural net, calculate the loss, from the loss calculate the gradients and from the gradient update the weights in our Neural Net properly.

In [27]:
import torch.nn.functional as F
# Performing Gradient Descent to improve the performance of the Network:
num_examples = xs.nelement()
print(num_examples)

for k in range(100):

    # Forward pass
    # We encode our input with the one-hot encoding, creating a tensor that is filled with 701 zeros and 1 one to signal which of the possible
    # 702 input combinations is currently being fed to the network.
    xenc = F.one_hot(xs, num_classes=702).float()
    
    logits = xenc @ W.T # Logits prediction
    counts = logits.exp() # Now we only have positive counts
    probs = counts / counts.sum(1, keepdim=True) # Now we have the probabilities outputted from the Neural Net for next character
    loss = -probs[torch.arange(num_examples), ys].log().mean() # The loss we use here is the Negative Log-likelihood Loss
    print(loss.item())

    # Backward pass
    W.grad = None # Set to zero the gradient
    loss.backward()

    # Update
    W.data += -60 * W.grad



156871
2.229524612426758
2.2288460731506348
2.22817325592041
2.227505683898926
2.22684383392334
2.226186990737915
2.2255356311798096
2.2248895168304443
2.2242486476898193
2.2236125469207764
2.2229814529418945
2.222355365753174
2.221734046936035
2.2211177349090576
2.220506429672241
2.2198996543884277
2.219297170639038
2.2186996936798096
2.218106746673584
2.2175183296203613
2.2169342041015625
2.2163546085357666
2.2157790660858154
2.215207815170288
2.2146413326263428
2.214078664779663
2.213519811630249
2.212965488433838
2.2124152183532715
2.21186900138855
2.2113265991210938
2.210787773132324
2.2102534770965576
2.2097229957580566
2.209196090698242
2.2086727619171143
2.208153009414673
2.2076375484466553
2.207125186920166
2.2066168785095215
2.2061116695404053
2.2056102752685547
2.2051119804382324
2.204617500305176
2.2041265964508057
2.2036385536193848
2.2031540870666504
2.2026729583740234
2.202195167541504
2.2017204761505127
2.20124888420105
2.2007808685302734
2.200315475463867
2.19985342025

Now let's evaluate the Neural Network on the Test-set:

In [22]:
# Creating the Test Dataset:

xs_test, ys_test = [], []

for w in test_names:
    chs = ['*'] + list(w) + ['*']
    for i in range(len(chs) - 2):
        doublet = chs[i] + chs[i + 1]
        ch2 = chs[i + 2]
        ix1 = doublettoi[doublet]
        ix2 = stoi[ch2]
        xs_test.append(ix1)
        ys_test.append(ix2)

xs_test = torch.tensor(xs_test)
ys_test = torch.tensor(ys_test)

num_examples = xs_test.nelement()
print('Number of examples: ', num_examples)

Number of examples:  39242


In [23]:
import torch.nn.functional as F
# Performing Gradient Descent to improve the performance of the Network:
num_examples = xs_test.nelement()
print(num_examples)



# Forward pass

# We encode our input with the one-hot encoding, creating a tensor that is filled with 701 zeros and 1 one to signal which of the possible
# 702 input combinations is currently being fed to the network.
xenc = F.one_hot(xs_test, num_classes=702).float()

logits = xenc @ W.T # Logits prediction
counts = logits.exp() # Now we only have positive counts
probs = counts / counts.sum(1, keepdim=True) # Now we have the probabilities outputted from the Neural Net for next character
loss = -probs[torch.arange(num_examples), ys_test].log().mean() # The loss we use here is the Negative Log-likelihood Loss
print(loss.item())

39242
2.2496397495269775


If you run the training code for a couple of times you can actually achieve lower loss on the test-set than approach 1. This is actually really interesting and I dont know quite why at this point. I did not think that this was possible actually. But my best loss for approach 1 was: 2.126 and loss for approach 2 was: 2.122.

Now we can simulate some names with approach 2 as well:

In [25]:
for i in range(20):


    first_char = torch.multinomial(first_character_probs, num_samples=1)
    first_char = itos[first_char.item()+1]
    first_doublet = f'*{first_char}'
    first_doublet_index = torch.tensor(doublettoi[first_doublet])

    start_symbol = F.one_hot(first_doublet_index, num_classes=702).float() # Inputting start symbol to the Neural Network to make it produce output characters
    current_doublet = start_symbol
    # Initializing current name:
    current_word = []
    # Adding the first randomly generated character:
    current_word.append(first_char)

    # This variable will keep track of the index of the current input doublet:
    next_doublet_index = torch.tensor(1)

    while(True):
        logits = current_doublet @ W.T # Logits prediction
        counts = logits.exp() # Now we only have positive counts
        probs = counts / counts.sum(0, keepdim=True)
        # probs = torch.ones(27) / 27 This line is for anyone curious about the difference between a Trigram and random generation

        # Sampling next character index:
        next_single_char_index = torch.multinomial(probs, num_samples=1)
        # If it is * we have reached the end of the name
        if next_single_char_index == 0:
            break

        # Plucking out next character from index:
        next_char = itos[next_single_char_index.item()]
        # Appending to current name:
        current_word.append(next_char)
        # Plucking out next doublet input:
        next_doublet = current_word[-2:]
        next_doublet = ''.join(next_doublet)
        next_doublet_index = torch.tensor(doublettoi[next_doublet])
        current_doublet = F.one_hot(next_doublet_index, num_classes=702).float()

    # Printing the current name:
    current_word = ''.join(current_word)
    print(current_word)

reia
aqvsxpummdyva
niyantcruvu
teryn
kamisilugs
asebz
jadlatra
marifpctjlporuccdvqnolia
sanor
vielikobi
tama
dan
beo
ty
mayaabzoodqyn
nyia
milynne
kath
ah
ran


## That's that!
Btw I'm really interested in business and AI.
For any business ideas please reach out to filiplarssonnnn@gmail.com

Bye!